services/gundeck/src/Gundeck/Push.hs

```diff
@@ -172,8 +188,10 @@ pushAll pushes = do
         resp <- compilePushResps targets <$> mpaBulkPush (compilePushReq <$> targets)
 
         -- native push
-        forM_ resp $ \((notif, psh), alreadySent) -> unless (psh ^. pushTransient) $
-            mpaPushNative notif psh =<< nativeTargets psh alreadySent
+        forM_ resp $ \((notif, psh), alreadySent) ->
+            unless (psh ^. pushTransient)
+                $ mpaRunWithBudget ()
+                    $ mpaPushNative notif psh =<< nativeTargets psh alreadySent
```

Comment:
Blocker: 
The budget call above increments the budget count by `1` each time this line is called. However, inside Native.push, there is a `mapConcurrently` on the amount of targets. If I send a message in a group chat with 300 participants each having 2 phones, budget is increased by 1 but 600 connections to SNS (worst case, if no one of those participants has the app and a websocket connection open) are opened at the same time. 

So setting a hard limit to X can mean vastly different amounts of connections to SNS are opened, depending on conversation size. This makes it hard to heuristically set a `hard` limit as setting it to `10` (random example number) means we would disallow 10 people in a 1-on-1 to send a message in the same milisecond (which we would not want as that's totally fine resource wise), but at the same time we do not want to allow 10 people sending a message to 600 devices in the same milisecond (as 6000 connections to SNS is probably too much).

Reply:
I was expecting this to be fine-grained enough, but we'll have to do experiments to find out who's right.

We could easily push down the calls to `runWithBudget`, but that would make it harder to make sure we haven't missed anything.

Alternatively, we could make `runWithBudget` take another argument that says by number the budget should be decreased, and compute that number purely from the context (`((notif, psh), alreadySent)` in this case.

Which one should we try next?

Reply:
My call would be to keep it as coarse-grained as it is now and see if that's enough.  It maybe, given that there is a wide enough gap between the budget needed for normal operation and what accumulates during outages.  But the extra parameter to `runWithBudget` doesn't sound difficult to implement either.

Reply:
I've implemented something.  You can be the judge of whether it's still clear enough to be useful.  I made an effort to make the individual commits easier to read than the whole change.